File: leetcode/1987. Number of Unique Good Subsequences.cpp
/* C++ code to find the number of unique good subsequences in a given binary string*/
#include <bits/stdc++.h>
using namespace std;

int numberOfUniqueGoodSubsequences(string binary)
{
    int mod = 1e9 + 7, endsWithZero = 0, endsWithOne = 0, hasZero = 0;
    for (int i = 0; i < binary.length(); ++i)
    {
        if (binary[i] == '1')
        {
            endsWithOne = (endsWithZero + endsWithOne + 1) % mod;
        }
        else
        {
            endsWithZero = (endsWithZero + endsWithOne) % mod;
            hasZero = 1;
        }
    }
    return (endsWithZero + endsWithOne + hasZero) % mod;
}

int main()
{
    string binary;
    binary = "1100100010101010100100000111110001111001000010000010010111011";
    cout << "The total number of unique good subsequences in the string "
        << binary << " are: "
        << numberOfUniqueGoodSubsequences(binary) << endl;
}

File: leetcode/1987. Number of Unique Good Subsequences.js
/**
 * @param {string} binary
 * @return {number}
 */
var numberOfUniqueGoodSubsequences = function (binary) {
  const len = binary.length;
  const mod = 1e9 + 7;
  /** @type {number} */
  let endsWithZero = 0, endsWithOne = 0;
  /** @type {boolean} */
  let hasZero = false;

  for (let i = 0; i < len; i++) {
    if (binary[i] === '1') {
      endsWithOne = (endsWithOne + endsWithZero + 1) % mod;
    } else {
      endsWithZero = (endsWithOne + endsWithZero) % mod;
      hasZero = true;
    }
  }
  return (endsWithOne + endsWithZero + (hasZero ? 1 : 0)) % mod;
};

console.log(
  "Number:",
  numberOfUniqueGoodSubsequences("1100100010101010100100000111110001111001000010000010010111011")
);